### PR TITLE
expression: fix wrong result when convert float to unsigned

### LINKED ISSUE #41736: An incorrect query result
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```SQL
CREATE TABLE t0(c0 tinyint(1) unsigned not null );
INSERT INTO t0 VALUES (1);
SELECT * FROM t0 WHERE CASE 0 WHEN (t0.c0)>(-1.194192591E9) THEN null ELSE 1 END; -- actual: {}, expected: {1}
```
### 2. What did you expect to see? (Required)
The query result should be {1}.
### 3. What did you see instead (Required)
The query result is an empty set.
### 4. What is your TiDB version? (Required)

Release Version: v6.6.0
Edition: Community
Git Commit Hash: f4ca0821fb96a2bdd37d2fb97eb26c07fc58d4e4
Git Branch: heads/refs/tags/v6.6.0
UTC Build Time: 2023-02-17 14:49:02
GoVersion: go1.19.5
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv

### PR DESCRIPTION
**Pull request by guo-shaoge**
### What problem does this PR solve?

Issue Number: close #41736

Problem Summary:
1. When compare `int column` with `non int constant`, will convert `non int constant` to int
2. [This logic](https://github.com/pingcap/tidb/blob/e6d1899492f1241926fb8baa411a7146effabf94/pkg/expression/builtin_compare.go#L1613) is used to check if underflow happens when convert to int, which requires we should set result as zero when convert X to unsigned int
3. But [ConvertFloatToUint](https://github.com/pingcap/tidb/blob/e6d1899492f1241926fb8baa411a7146effabf94/pkg/types/convert.go#L172) will not obey this requirement when `AllowNegativeToUnsigned` is disabled.

PS: what does `AllowNegativeToUnsigned` mean, check description [here](https://github.com/pingcap/tidb/blob/57d0b40cd9610351a6769926e5c9e30ddd5f6b7c/pkg/executor/executor.go#L2008)

### What changed and how does it work?
Disable `AllowNegativeToUnsigned` when conversion happens when compare `int column` with `non int constant`.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
expression: fix wrong result when convert float to unsigned
```

### CODE REVIEW

File: pkg/expression/integration_test/integration_test.go [hunk 1]
```diff
@@ -3032,3 +3032,14 @@ func TestIssue43527(t *testing.T) {
 		"SELECT @total := @total + d FROM (SELECT d FROM test) AS temp, (SELECT @total := b FROM test) AS T1 where @total >= 100",
 	).Check(testkit.Rows("200", "300", "400", "500"))
 }
+
```
Comment (guo-shaoge) [inline pkg/expression/integration_test/integration_test.go:3035]: todo add mysqltest instead here.

**Review by SeaRise — APPROVED** [review]
Sadly, this bug also appeared on tiflash

**Comment by guo-shaoge** [comment]
> Sadly, this bug also appeared on tiflash

After this fix, tiflash will be ok. Because this fix happens plan generation, so tilflash will receive correct plan after this fix.

**Review by wshwsh12 — APPROVED** [review]

**Review by windtalker — APPROVED** [review]
LGTM

**Comment by MimeLyc** [comment]
update

**Pull request merged**